hedera-node/src/main/java/com/hedera/services/utils/MiscUtils.java

```diff
@@ -411,260 +514,116 @@ public static JKey asFcKeyUnchecked(Key key) {
 			return "NotImplemented";
 		}
 	}
 
-	public static HederaFunctionality functionOf(TransactionBody txn) throws UnknownHederaFunctionality {
-		if (txn.hasSystemDelete()) {
-			return SystemDelete;
-		} else if (txn.hasSystemUndelete()) {
-			return SystemUndelete;
-		} else if (txn.hasContractCall()) {
-			return ContractCall;
-		} else if (txn.hasContractCreateInstance()) {
-			return ContractCreate;
-		} else if (txn.hasContractUpdateInstance()) {
-			return ContractUpdate;
-		} else if (txn.hasCryptoAddLiveHash()) {
-			return CryptoAddLiveHash;
-		} else if (txn.hasCryptoCreateAccount()) {
-			return CryptoCreate;
-		} else if (txn.hasCryptoDelete()) {
-			return CryptoDelete;
-		} else if (txn.hasCryptoDeleteLiveHash()) {
-			return CryptoDeleteLiveHash;
-		} else if (txn.hasCryptoTransfer()) {
-			return CryptoTransfer;
-		} else if (txn.hasCryptoUpdateAccount()) {
-			return CryptoUpdate;
-		} else if (txn.hasFileAppend()) {
-			return FileAppend;
-		} else if (txn.hasFileCreate()) {
-			return FileCreate;
-		} else if (txn.hasFileDelete()) {
-			return FileDelete;
-		} else if (txn.hasFileUpdate()) {
-			return FileUpdate;
-		} else if (txn.hasContractDeleteInstance()) {
-			return ContractDelete;
-		} else if (txn.hasFreeze()) {
-			return Freeze;
-		} else if (txn.hasConsensusCreateTopic()) {
-			return ConsensusCreateTopic;
-		} else if (txn.hasConsensusUpdateTopic()) {
-			return ConsensusUpdateTopic;
-		} else if (txn.hasConsensusDeleteTopic()) {
-			return ConsensusDeleteTopic;
-		} else if (txn.hasConsensusSubmitMessage()) {
-			return ConsensusSubmitMessage;
-		} else if (txn.hasTokenCreation()) {
-			return TokenCreate;
-		} else if (txn.hasTokenFreeze()) {
-			return TokenFreezeAccount;
-		} else if (txn.hasTokenUnfreeze()) {
-			return TokenUnfreezeAccount;
-		} else if (txn.hasTokenGrantKyc()) {
-			return TokenGrantKycToAccount;
-		} else if (txn.hasTokenRevokeKyc()) {
-			return TokenRevokeKycFromAccount;
-		} else if (txn.hasTokenDeletion()) {
-			return TokenDelete;
-		} else if (txn.hasTokenUpdate()) {
-			return TokenUpdate;
-		} else if (txn.hasTokenMint()) {
-			return TokenMint;
-		} else if (txn.hasTokenBurn()) {
-			return TokenBurn;
-		} else if (txn.hasTokenWipe()) {
-			return TokenAccountWipe;
-		} else if (txn.hasTokenAssociate()) {
-			return TokenAssociateToAccount;
-		} else if (txn.hasTokenDissociate()) {
-			return TokenDissociateFromAccount;
-		} else if (txn.hasTokenFeeScheduleUpdate()) {
-			return TokenFeeScheduleUpdate;
-		} else if (txn.hasScheduleCreate()) {
-			return ScheduleCreate;
-		} else if (txn.hasScheduleSign()) {
-			return ScheduleSign;
-		} else if (txn.hasScheduleDelete()) {
-			return ScheduleDelete;
-		} else if (txn.hasUncheckedSubmit()) {
-			return UncheckedSubmit;
-		} else {
-			throw new UnknownHederaFunctionality();
+	public static HederaFunctionality functionOf(final TransactionBody txn) throws UnknownHederaFunctionality {
```

Comment:
As above, the performance hit will be **massive** here; and since this is used in `handleTransaction()` I'd definitely revert to the old code.

Reply:
I was thinking of performance. That was why I had all of these functions initialized statically (only once upfront). I think the invoke should be the same as calling the function directly by name. The only difference is that we have to retrieve which function to call via a hashmap which is supposed to be equally cheap.

Reply:
@tinker-michaelj, these are crypto transfer and hcs 10k tests for the new code.
https://hedera-hashgraph.slack.com/archives/C018Y4E6ADT/p1629523640050800
I don't see any sign of regression. I will leave it run daily over the weekend for more data points.

Reply:
> @tinker-michaelj, these are crypto transfer and hcs 10k tests for the new code.
> https://hedera-hashgraph.slack.com/archives/C018Y4E6ADT/p1629523640050800
> I don't see any sign of regression. I will leave it run daily over the weekend for more data points.

System-level tests are too rough-grained to evaluate this kind of impact.

I did a quick "old vs. new" benchmark in a branch [here](https://github.com/hashgraph/hedera-services/blob/unit-test-functionOf-cmp/hedera-node/src/test/java/com/hedera/services/utils/MiscUtilsTest.java#L666), and measured a 41x slowdown with the new code:
```
Done classifying 380000000 transaction with old vs. new code
 -> Time w/ old code: 134ms
 -> Time w/ new code: 5564ms (41x slower)
```

This is not even the biggest problem though, since at 10k TPS of transfers the new `MiscUtils.functionOf()` implementation will be creating ~30,000 objects/sec that we don't need...
```
			for (var f : SCHEDULE_FUNCTIONS) { <<-- This creates an iterator
				if ((boolean) TRANSACTION_HAS_METHODS.get(TRANSACTION_FUNCTION_TO_STRINGS.get(f)).invoke(txn)) {
					return Pair.of(f, "");
				}
			}
			for (var f : sameNameNonScheduleFunctions) { <<-- This creates an iterator
				if ((boolean) TRANSACTION_HAS_METHODS.get(f.name()).invoke(txn)) {
					return Pair.of(f, ""); <<--This creates a Pair
				}
			}
```
Especially in the `handleTransaction` path, there's no reason to accept unnecessary costs like these without a huge improvement in the simplicity of the code.